examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py

```diff
@@ -2192,13 +2192,12 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
                 # run inference
                 generator = torch.Generator(device=accelerator.device).manual_seed(args.seed) if args.seed else None
                 pipeline_args = {"prompt": args.validation_prompt}
-                inference_ctx = (
-                    contextlib.nullcontext()
-                    if "playground" in args.pretrained_model_name_or_path
-                    else torch.cuda.amp.autocast()
-                )
+                if torch.backends.mps.is_available() or "playground" in args.pretrained_model_name_or_path:
```

Comment:
Autocast isn't supported for Playground models?

Reply:
Nope. The validation runs go crazy when doing autocasting for the Playground model. More details: https://github.com/huggingface/diffusers/pull/7126

Reply:
autocast in general seems like a huge mistake to rely on.

last year the advice from patrick von platen was to never use it for inference, and here we are, abusing the heck out of it at inference time

Reply:
Fair to be skeptical. But I am really hesitant to touch what's been working for the majority of our users (GPU training i.e.,) for a while now. 

>  abusing the heck out of it at inference time

Little err. Only during intermediate validation runs during training.

Reply:
did you know autocast when running training in bf16, will ruin every tensor by casting them down to fp16 and then back up to bf16? :D

Reply:
Nope. Cannot possibly exhaustively know everything beforehand.

Reply:
> Nope. Cannot possibly exhaustively know everything beforehand.

Something very bad is happening with autocast. See my replies in this issue here: https://github.com/pytorch/pytorch/issues/120376

Basically bfloat16 performance should be matching float32 and I rewrote the AdamW optimiser to do so, but you can see in the graph that when you store the weights in float32 and then autocast to bfloat16 training completely breaks.

I believe this problem also affects all accelerate scripts that commonly store weights in float32 and then train in bfloat16. And with the improved optimiser, everyone should really just be storing weights in bfloat16 and training with bfloat16 autocast anyway. We really need to get torch in on this to fix the autocast feature, because something appears to be going very wrong.

Reply:
i am in agreement. i had autocast disabled in simpletuner for about a year and now that it's back in with my changes this week, things have become a nightmare for reliability

Reply:
Kenji on the LAION Discord was able to figure out how to fix autocast with float32 weights and bfloat16, basically autocast is dangerous to use outside of the loss calculate and the `forward` step as it will completely corrupt the optimisation step. Once you put it there (and ONLY there) the optimiser will work in float32 and everything should be fixed.

![Screenshot_2024-04-02_09-22-08](https://github.com/huggingface/diffusers/assets/110263573/1161f723-3b59-4b3e-ae80-d66ae0588ea4)

That said if we have a good bf16 version of AdamW now, it makes sense to save on the weights and just use that.